lib/SILGen/SILGenPoly.cpp

```diff
@@ -3407,3 +3407,113 @@ void SILGenFunction::emitProtocolWitness(Type selfType,
   scope.pop();
   B.createReturn(loc, reqtResultValue);
 }
+
+//===----------------------------------------------------------------------===//
+// Conversion to Canonical SILFunctionType Thunks
+//===----------------------------------------------------------------------===//
+
+static void buildCanonicalFunctionThunkBody(SILGenFunction &SGF,
+                                            SILLocation loc,
+                                            CanSILFunctionType nonCanonicalTy,
+                                            CanSILFunctionType canonicalTy) {
+  SGF.F.setBare(IsBare);
+  SGF.F.setThunk(IsThunk);
+
+  FullExpr scope(SGF.Cleanups, CleanupLocation::get(loc));
+  FormalEvaluationScope formalEvalScope(SGF);
+
+  // Collect the thunk params, creating arguments for each parameter.
+  SmallVector<ManagedValue, 8> origParams;
+  SGF.collectThunkParams(loc, origParams);
+
+  AbstractionPattern canonicalTyPattern(canonicalTy);
+  AbstractionPattern nonCanonicalTyPattern(nonCanonicalTy);
+
+  // Then translate our parameters into new params.
+  SmallVector<ManagedValue, 8> newParams;
+  TranslateArguments(SGF, loc, llvm::makeArrayRef(origParams).drop_back(1),
+                     newParams, nonCanonicalTy->getParameters())
+      .translate(canonicalTyPattern, canonicalTy, nonCanonicalTyPattern,
```

Comment:
translate() takes AST types not SIL function types, so this won't work

Reply:
So I should just write my own?

Reply:
Ok. I'll just write my own.

Reply:
I don't think it has to do much -- just retain a value if it comes in at +0 and is taken at +1?